### PR TITLE
add VMAS scale up/down support for openshift

### PR DESCRIPTION
**Pull request by jim-minter**
adds VMAS scale up/down support for openshift

`acs-engine scale --resource-group $RG --location $LOCATION --subscription-id $SUB --new-node-count $COUNT --deployment-dir _output/$RG --node-pool $POOL --master-FQDN https://$RG.$LOCATION.cloudapp.azure.com:8443/`

### CODE REVIEW

**Comment by jim-minter** [comment]
@kargakis @pweil- ptal

@kargakis, do you think this functionality can/should be integrated into e2e tests? any thoughts on how?

@kargakis assuming we go ahead persisting the bundles in the apimodel it should be possible to have the e2e test retrieve the kubeconfig from there -- shall I open a follow-up issue to track that?

File: cmd/scale.go [hunk 1]
```diff
@@ -320,6 +345,15 @@ func (sc *scaleCmd) run(cmd *cobra.Command, args []string) error {
 	addValue(parametersJSON, sc.agentPool.Name+"Count", countForTemplate)
 
 	switch orchestratorInfo.OrchestratorType {
+	case api.OpenShift:
+		err = transformer.NormalizeForOpenShiftVMASScalingUp(sc.logger, sc.agentPool.Name, templateJSON)
+		if err != nil {
+			log.Fatalf("error tranforming the template for scaling template %s: %s", sc.apiModelPath, err.Error())
+			os.Exit(1)
```
Comment (kargakis) [inline cmd/scale.go]: Fatal already os.Exits

File: cmd/scale.go [hunk 2]
```diff
@@ -237,8 +239,31 @@ func (sc *scaleCmd) run(cmd *cobra.Command, args []string) error {
...
 			}
 
-			if orchestratorInfo.OrchestratorType == api.Kubernetes {
-				err = sc.drainNodes(vmsToDelete)
+			switch orchestratorInfo.OrchestratorType {
+			case api.Kubernetes:
+				kubeConfig, err := acsengine.GenerateKubeConfig(sc.containerService.Properties, sc.location)
+				if err != nil {
+					log.Errorf("failed to generate kube config: %v", err)
+					return err
```
Comment (kargakis) [inline cmd/scale.go]: Logging an error and returning it seems redundant. What is the caller doing with it? Maybe just 
```go
return fmt.Errorf("failed to generate kube config: %v", err)
```

File: cmd/scale.go [hunk 3]
```diff
@@ -237,8 +239,31 @@ func (sc *scaleCmd) run(cmd *cobra.Command, args []string) error {
...
+				}
+			case api.OpenShift:
+				bundle := bytes.NewReader(sc.containerService.Properties.OrchestratorProfile.OpenShiftConfig.ConfigBundles["master"])
+				fs, err := filesystem.NewTGZReader(bundle)
+				if err != nil {
+					log.Errorf("failed to read master bundle: %v", err)
+					return err
+				}
+				kubeConfig, err := fs.ReadFile("etc/origin/master/admin.kubeconfig")
```
Comment (kargakis) [inline cmd/scale.go]: FWIW, In Kubernetes it seems they are generating a new kubeconfig so they are not reusing the one that was generated during cluster startup.

File: pkg/acsengine/transform/transform.go [hunk 4]
```diff
@@ -97,6 +99,54 @@ func (t *Transformer) NormalizeForVMSSScaling(logger *logrus.Entry, templateMap
 	return nil
 }
 
+// NormalizeForOpenShiftVMASScalingUp takes a template and removes elements that are unwanted in a K8s VMAS scale up/down case
```
Comment (kargakis) [inline pkg/acsengine/transform/transform.go]: s/K8s/OpenShift/

File: pkg/acsengine/transform/transform.go [hunk 5]
```diff
@@ -97,6 +99,54 @@ func (t *Transformer) NormalizeForVMSSScaling(logger *logrus.Entry, templateMap
...
+func (t *Transformer) NormalizeForOpenShiftVMASScalingUp(logger *logrus.Entry, agentPoolName string, templateMap map[string]interface{}) (err error) {
```
Comment (kargakis) [inline pkg/acsengine/transform/transform.go]: Please add a unit test for this

**Comment by kargakis** [comment]
> @kargakis assuming we go ahead persisting the bundles in the apimodel it should be possible to have the e2e test retrieve the kubeconfig from there -- shall I open a follow-up issue to track that?

How is this going to look like when we allow users bring their own certificates? Maybe we can populate the certificate profile based on the config bundle?

**Comment by kargakis** [comment]
> Maybe we can populate the certificate profile based on the config bundle?

In case we generate.

**Comment by pweil-** [comment]
> do you think this functionality can/should be integrated into e2e tests?

Yes to the should question.  If not here then I want it in our OpenShift testing.

File: pkg/api/vlabs/validate.go [hunk 6]
```diff
@@ -280,7 +280,10 @@ func validateImageNameAndGroup(name, resourceGroup string) error {
 }
 
 // Validate implements APIObject
-func (m *MasterProfile) Validate() error {
+func (m *MasterProfile) Validate(o *OrchestratorProfile) error {
+	if o.OrchestratorType == OpenShift && m.Count != 1 {
```
Comment (pweil-) [inline pkg/api/vlabs/validate.go]: unit test (reading commit by commit, if it comes later than ignore :smile:)

File: cmd/scale.go [hunk 7]
```diff
@@ -237,8 +239,31 @@ func (sc *scaleCmd) run(cmd *cobra.Command, args []string) error {
...
+				if err != nil {
+					log.Errorf("failed to read kube config: %v", err)
+					return err
+				}
+				err = sc.drainNodes(string(kubeConfig), vmsToDelete)
```
Comment (pweil-) [inline cmd/scale.go]: the drain node calls can be removed from the case statement, the only difference to account for is getting the kube config to pass in.

File: parts/openshift/openshiftnodescript.sh [hunk 8]
```diff
@@ -23,3 +23,8 @@ update-ca-trust
 # note: ${SERVICE_TYPE}-node crash loops until master is up
 systemctl enable ${SERVICE_TYPE}-node.service
 systemctl start ${SERVICE_TYPE}-node.service &
+
+while [[ $(KUBECONFIG=/etc/origin/node/node.kubeconfig oc get node $(hostname) -o template \
```
Comment (pweil-) [inline parts/openshift/openshiftnodescript.sh]: It's ok for a node to be unready as the system comes up, why wait?

File: pkg/acsengine/defaults.go [hunk 9]
```diff
@@ -740,6 +740,11 @@ func setStorageDefaults(a *api.Properties) {
 }
 
 func openShiftSetDefaultCerts(a *api.Properties) (bool, error) {
+	if len(a.OrchestratorProfile.OpenShiftConfig.ConfigBundles["master"]) > 0 &&
```
Comment (pweil-) [inline pkg/acsengine/defaults.go]: wish there was a clearer indicator that this is a non-generate command

**Comment by CecileRobertMichon** [comment]
> I can't find any scale tests in circleci e2e tests. Unclear whether there are any that run out of bound in Jenkins @jackfrancis @CecileRobertMichon

Yes, our scale and upgrade tests run on Jenkins. Let us know which apimodel definition you would like to test and we can set up a new job.

File: cmd/scale.go (hunk 3 above)
Comment (jim-minter) [reply cmd/scale.go]: Yes, but in the k8s case the actual authentication certificate within the kubeconfig is still the originally generated one.  While not perfect I think our approach is an improvement.

File: cmd/scale.go (hunk 2 above)
Comment (jim-minter) [reply cmd/scale.go]: This is a code move for k8s and I'm reluctant to refactor.  I'll ensure our newly added code is cleaner in this regard.  https://github.com/Azure/acs-engine/issues/2913 for the other code.

File: cmd/scale.go (hunk 1 above)
Comment (jim-minter) [reply cmd/scale.go]: the whole file needs a cleanup - I'm a bit torn between A) matching existing code, B) adding cleaner code, C) trying to improve the whole thing.  I'll move my stance from A) to B), how about that?

File: cmd/scale.go (hunk 7 above)
Comment (jim-minter) [reply cmd/scale.go]: however the drain node calls still must only run when the orchestrator is k8s or ocp - that's why they're in the switch still

File: parts/openshift/openshiftnodescript.sh (hunk 8 above)
Comment (jim-minter) [reply parts/openshift/openshiftnodescript.sh]: it has the advantage that the scale up call doesn't return success until all the requested nodes have successfully joined the cluster, and that if any new node fails to register, the error is reported as a deployment failure

File: pkg/acsengine/defaults.go (hunk 9 above)
Comment (jim-minter) [reply pkg/acsengine/defaults.go]: I think the key here is "Default"

**Comment by jim-minter** [comment]
reviews addressed, including adding unit tests (thanks! :grimacing:)
need to decide what to do about e2e testing.

**Comment by jim-minter** [comment]
> Yes, our scale and upgrade tests run on Jenkins. Let us know which apimodel definition you would like to test and we can set up a new job.

Wondering if we can/should do this in the e2e so we can make additional checks (e.g. that the right node labels are applied after scale up)?

**Comment by CecileRobertMichon** [comment]
> Wondering if we can/should do this in the e2e so we can make additional checks (e.g. that the right node labels are applied after scale up)?

@jim-minter our Jenkins jobs run e2e tests before and after scaling the cluster so we do make those checks. Rationale is to avoid slowing down circleci with scale and upgrade since they take a long time to test (we run multiple scale operations on the same cluster and check it's healthy each time with e2e)

**Comment by jim-minter** [comment]
Have added to the e2e test to ensure that created nodes are labeled correctly, plus 1 more commit with non-controversial fixes to the e2e tests

File: cmd/scale.go (hunk 2 above)
Comment (jackfrancis) [reply cmd/scale.go]: Thanks for filing the backlog item ❤️

File: pkg/api/vlabs/validate_test.go [hunk 10]
```diff
@@ -829,6 +829,41 @@ func TestValidateImageNameAndGroup(t *testing.T) {
...
+		expectedErr      bool
+	}{
+		{
+			orchestratorType: OpenShift,
+			count:            1,
+		},
+		{
+			orchestratorType: OpenShift,
+			count:            3,
+			expectedErr:      true,
```
Comment (pweil-) [inline pkg/api/vlabs/validate_test.go]: tests need to test against a known, expected error in failure cases to really know you're hitting the code you think you are.  I'd also make the test provide the `MasterProfile` so this really can be used as a  `TestMasterProfileValidate` in the future and not just a count test.

File: test/e2e/openshift/openshift_test.go [hunk 11]
```diff
@@ -56,6 +56,32 @@ var _ = Describe("Azure Container Cluster using the OpenShift Orchestrator", fun
 		Expect(ready).To(Equal(true))
 	})
 
+	It("should label nodes correctly", func() {
```
Comment (pweil-) [inline test/e2e/openshift/openshift_test.go]: :heart:

File: pkg/api/vlabs/validate_test.go (hunk 10 above)
Comment (kargakis) [reply pkg/api/vlabs/validate_test.go]: agreed

File: parts/k8s/kubernetesparams.t [hunk 12]
```diff
@@ -494,7 +494,7 @@
       "type": "string"
     },
     "kubernetesClusterAutoscalerEnabled": {
-      {{PopulateClassicModeDefaultValue "kubernetesClusterAutoscalerEnabled"}}
+      "defaultValue": false,
```
Comment (kargakis) [inline parts/k8s/kubernetesparams.t]: Fix appropriately or should we fix this in a separate PR?

Comment (kargakis) [reply parts/k8s/kubernetesparams.t]: Seems to be fixed in https://github.com/Azure/acs-engine/pull/2915, ignore

**Comment by jim-minter** [comment]
@pweil- updated - ptal

current thinking after discussion with @jackfrancis is that for now we'll instigate a periodic test on our side which proves that the scale up use case works to our satisfaction - trello card opened for this

**Comment by jim-minter** [comment]
gofmt error fixed, thanks

@jackfrancis or @CecileRobertMichon please merge

File: parts/openshift/openshiftnodescript.sh (hunk 8 above)
Comment (kargakis) [reply parts/openshift/openshiftnodescript.sh]: Is there any timeout used during scale up?

Comment (jim-minter) [reply parts/openshift/openshiftnodescript.sh]: same as on deploy - arm api will timeout, not sure exactly after how long - something like 45 minutes?

**Pull request merged**